byterun/sys.c

```diff
@@ -368,6 +369,31 @@ CAMLprim value caml_sys_get_argv(value unit)
 
 void caml_sys_init(char_os * exe_name, char_os **argv)
 {
+#ifdef _WIN32
+  /* Determine the version of Windows we're running, and cache it */
+  WCHAR fileName[MAX_PATH];
+  DWORD size =
+    GetModuleFileName(GetModuleHandle(L"kernel32"), fileName, MAX_PATH);
+  DWORD dwHandle = 0;
+  BYTE* versionInfo;
+  fileName[size] = 0;
+  size = GetFileVersionInfoSize(fileName, &dwHandle);
+  versionInfo = (BYTE*)malloc(size * sizeof(BYTE));
+  if (GetFileVersionInfo(fileName, 0, size, versionInfo)) {
+    UINT len = 0;
+    VS_FIXEDFILEINFO* vsfi = NULL;
+    VerQueryValue(versionInfo, L"\\", (void**)&vsfi, &len);
+    caml_win32_major = HIWORD(vsfi->dwFileVersionMS);
+    caml_win32_minor = LOWORD(vsfi->dwFileVersionMS);
+    caml_win32_build = HIWORD(vsfi->dwFileVersionLS);
+    caml_win32_revision = LOWORD(vsfi->dwFileVersionLS);
+  }
+  free(versionInfo);
+
+#if WINDOWS_UNICODE
+  SetConsoleOutputCP(CP_UTF8);
+#endif
+#endif
```

Comment:
Again, this looks like a break of abstraction levels to me. Could this be moved to its own function?

Reply:
Initialising the system in a function called `caml_sys_init` seemed vaguely appropriate to me?! I freely admit that I selected on the basis that it is called from all of the various entry points for starting an OCaml application!

Reply:
I was thinking that the logic to get the win32 version could be moved into a function called something like `caml_win32_init_version` that could be called from `caml_sys_init` (the OutputCP thing is of course a separate concern). The idea is, from the point of view of someone looking at `caml_sys_init` and interested about a high-level overview of what is happening here, to not have to read through 19 lines of low-level Windows-specific code.